hilti/toolchain/src/compiler/parser/parser.yy

```diff
@@ -508,10 +508,10 @@ stmt          : stmt_expr ';'                    { $$ = std::move($1); }
               | TRY block try_catches
                                                  { $$ = builder->statementTry(std::move($2), std::move($3), __loc__); }
               | ASSERT expr ';'                  { $$ = builder->statementAssert(std::move($2), {}, __loc__); }
-              | ASSERT expr ':' expr ';'         { $$ = builder->statementAssert(std::move($2), std::move($4), __loc__); }
-              | ASSERT_EXCEPTION expr ';'        { $$ = builder->statementAssert(hilti::statement::assert::Exception(), std::move($2), {}, {}, __loc__); }
-              | ASSERT_EXCEPTION expr ':' expr ';'
+              | ASSERT_EXCEPTION expr_no_or_error ':' expr ';'
                                                  { $$ = builder->statementAssert(hilti::statement::assert::Exception(), std::move($2), {}, std::move($4), __loc__); }
+              | ASSERT_EXCEPTION expr_no_or_error ';'
```

Comment:
We could still have `expr_no_or_error` contain a `result<void>` and by using this constructor we would report an incorrect error, e.g.,

```ruby
global x = 0 : "foo";
assert x;
```

```console
$ spicyc -j foo.spicy
uncaught exception hilti::rt::AssertionFailure: failed expression 'x' (foo.spicy:15:1-15:9)
```

Part of this comes from the fact that `Assert` can hold either an expression and a message, or a `result<void>` and no message. To use `Assert::message` one needs to check the type of the expression, but this would also be error prone like e.g., with this patch

```ruby
global x = result("foo");
# Should work since a result did convert implicitly to a bool.
assert x : "but I initialized it!"; # ERROR: cannot coerce expression 'x' of type 'string' to type 'bool'
```

I think this needs cleanup in `Assert`.

Reply:
> global x = 0 : "foo";
> assert x;
> 
> $ spicyc -j foo.spicy
> uncaught exception hilti::rt::AssertionFailure: failed expression 'x' (foo.spicy:15:1-15:9)

This seems correct, no? `0 : "foo"` gets the error condition because 0 coerces to false. And then the assert does a boolean check on its `result<void>` argument and finds it false, hence it triggers.

> global x = result("foo");
> assert x : "but I initialized it!"; # ERROR: cannot coerce expression 'x' of type 'string' to type 'bool'

The bug here seems elsewhere: `x` ends up having a type of string. If I change the first line to `global x: result<string> = result("foo");` it works. Will look into that.

Reply:
> This seems correct, no?

Or are you saying it should print the message?

Reply:
> Or are you saying it should print the message?

Ah, and if so, it's actually the same issue as the 2nd. This works:

```
global x: result<void> = 0 : "foo";
assert x;
...
uncaught exception hilti::rt::AssertionFailure: foo (b.spicy:3:1-3:9)
```